runtime/compiler/optimizer/J9RecognizedCallTransformer.cpp

```diff
@@ -102,26 +103,211 @@ void J9::RecognizedCallTransformer::process_java_lang_StringUTF16_toBytes(TR::Tr
-   node->removeChild(0);
-   bool safeToSkipDiamond = !strncmp(comp()->getCurrentMethod()->classNameChars(), "java/util/concurrent/atomic/", strlen("java/util/concurrent/atomic/"));
+   bool isNotStaticField = !strncmp(comp()->getCurrentMethod()->classNameChars(), "java/util/concurrent/atomic/", strlen("java/util/concurrent/atomic/"));
 
-   TR_Array<TR::Node*> params(trMemory(), node->getNumChildren()-2);
-   for (int i = 2; i < node->getNumChildren(); i++)
+   TR::Node* unsafeCall = treetop->getNode()->getFirstChild();
+   TR::Node* objectNode = unsafeCall->getChild(1);
+   TR::Node* offsetNode = unsafeCall->getChild(2);
+   TR::Node* address = NULL;
+
+   // Preserve null check on the unsafe object
+   if (treetop->getNode()->getOpCode().isNullCheck())
       {
-      params.add(TR::TransformUtil::saveNodeToTempSlot(comp(), node->getChild(i), treetop));
+      TR::Node *passthrough = TR::Node::create(unsafeCall, TR::PassThrough, 1);
+      passthrough->setAndIncChild(0, unsafeCall->getFirstChild());
+      TR::Node * checkNode = TR::Node::createWithSymRef(treetop->getNode(), TR::NULLCHK, 1, passthrough, treetop->getNode()->getSymbolReference());
+      treetop->insertBefore(TR::TreeTop::create(comp(), checkNode));
+      TR::Node::recreate(treetop->getNode(), TR::treetop);
       }
 
-   auto address = TR::TransformUtil::calculateUnsafeAddress(treetop, node->getChild(0), node->getChild(1), comp(), false, safeToSkipDiamond);
-   node->removeAllChildren();
-   node->setNumChildren(params.size()+1);
-   node->setAndIncChild(0, address);
-   for (int i = 0; i < params.size(); i++)
+   if (isNotStaticField)
+      {
+      // It is safe to skip diamond, the address can be calaulated directly via [object+offset]
+      address = TR::Compiler->target.is32Bit() ? TR::Node::create(TR::aiadd, 2, objectNode, TR::Node::create(TR::l2i, 1, offsetNode)) :
+                                              TR::Node::create(TR::aladd, 2, objectNode, offsetNode);
+      }
+   else
       {
-      node->setAndIncChild(i+1, params[i]);
+      // Otherwise, the address is [object+offset] for non-static field,
+      //            or [object's ramStaticsFromClass + (offset & ~mask)] for static field
+
+      // Save all the children to temps before splitting the block
+      TR::TransformUtil::createTempsForCall(this, treetop);
+
+      auto cfg = comp()->getMethodSymbol()->getFlowGraph();
+      objectNode = unsafeCall->getChild(1);
+      offsetNode = unsafeCall->getChild(2);
+
+      // Null Check
+      if (needsNullCheck)
+         {
+         auto NULLCHKNode = TR::Node::createWithSymRef(TR::NULLCHK, 1, 1,
+                                                       TR::Node::create(TR::PassThrough, 1, objectNode->duplicateTree()),
+                                                       comp()->getSymRefTab()->findOrCreateNullCheckSymbolRef(comp()->getMethodSymbol()));
+         NULLCHKNode->getByteCodeInfo().setCallerIndex(comp()->getCurrentInlinedSiteIndex());
```

Comment:
what bytecode offset do we give this? it should match the original unsafe call in case we have to yield I think...

Reply:
I passed `unsafeCall` as the originating node to `createWithSymRef`. The null check node should have the bytecode info of the `unsafeCall` now.